core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -425,17 +369,21 @@ private UpdateResultHolder executeUpdateRequest(UpdateRequest updateRequest, Ind
      * Execute the given {@link IndexRequest} on a replica shard, throwing a
      * {@link RetryOnReplicaException} if the operation needs to be re-tried.
      */
-    public static Engine.IndexResult executeIndexRequestOnReplica(IndexRequest request, IndexShard replica) throws IOException {
+    public static Engine.IndexResult executeIndexRequestOnReplica(DocWriteResponse primaryResponse, IndexRequest request, IndexShard replica) throws IOException {
         final ShardId shardId = replica.shardId();
         SourceToParse sourceToParse =
             SourceToParse.source(SourceToParse.Origin.REPLICA, shardId.getIndexName(), request.type(), request.id(), request.source(),
                 request.getContentType()).routing(request.routing()).parent(request.parent());
 
         final Engine.Index operation;
+        final long version = primaryResponse.getVersion();
+        final VersionType versionType = request.versionType().versionTypeForReplicationAndRecovery();
```

Comment:
Sadly, that doesn't work as the version type can be FORCE (albeit deprecated) or EXTERNAL_GTE . That's why had to make that crazy assertion in https://github.com/elastic/elasticsearch/pull/23167